fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -0,0 +1,373 @@
+#include "fdbclient/SpecialKeySpace.actor.h"
+#include "flow/UnitTest.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+// This function will normalize the given KeySelector to a standard KeySelector:
+// orEqual == false && offset == 1 (Standard form)
+// If the corresponding key is not in this special key range, it will move as far as possible to adjust the offset to 1
+// It does have overhead here since we query all keys twice in the worst case.
+// However, moving the KeySelector while handling other parameters like limits makes the code much more complex and hard
+// to maintain Separate each part to make the code easy to understand and more compact
+ACTOR Future<Void> SpecialKeyRangeBaseImpl::normalizeKeySelectorActor(const SpecialKeyRangeBaseImpl* pkrImpl,
+                                                                      Reference<ReadYourWritesTransaction> ryw,
+                                                                      KeySelector* ks) {
+	ASSERT(!ks->orEqual); // should be removed before calling
+	ASSERT(ks->offset != 1); // never being called if KeySelector is already normalized
+
+	state Key startKey(pkrImpl->range.begin);
+	state Key endKey(pkrImpl->range.end);
+
+	if (ks->offset < 1) {
+		// less than the given key
+		if (pkrImpl->range.contains(ks->getKey())) endKey = keyAfter(ks->getKey());
+	} else {
+		// greater than the given key
+		if (pkrImpl->range.contains(ks->getKey())) startKey = ks->getKey();
+	}
+
+	TraceEvent("NormalizeKeySelector")
+	    .detail("OriginalKey", ks->getKey())
+	    .detail("OriginalOffset", ks->offset)
+	    .detail("SpecialKeyRangeStart", pkrImpl->range.begin)
+	    .detail("SpecialKeyRangeEnd", pkrImpl->range.end);
+
+	Standalone<RangeResultRef> result = wait(pkrImpl->getRange(ryw, KeyRangeRef(startKey, endKey)));
+	if (result.size() == 0) {
+		TraceEvent("ZeroElementsIntheRange").detail("Start", startKey).detail("End", endKey);
+		return Void();
+	}
+	// Note : KeySelector::setKey has byte limit according to the knobs, customize it if needed
+	if (ks->offset < 1) {
+		if (result.size() >= 1 - ks->offset) {
+			ks->setKey(KeyRef(ks->arena(), result[result.size() - (1 - ks->offset)].key));
+			ks->offset = 1;
+		} else {
+			ks->setKey(KeyRef(ks->arena(), result[0].key));
+			ks->offset += result.size();
+		}
+	} else {
+		if (result.size() >= ks->offset) {
+			ks->setKey(KeyRef(ks->arena(), result[ks->offset - 1].key));
+			ks->offset = 1;
+		} else {
+			ks->setKey(KeyRef(ks->arena(), keyAfter(result[result.size() - 1].key)));
+			ks->offset -= result.size();
+		}
+	}
+	TraceEvent("NormalizeKeySelector")
```

Comment:
This might be too verbose, let's make this SevDebug for now

Reply:
Done. Thanks.